src/pycolmap/pybind11_extension.h

```diff
@@ -13,10 +13,53 @@
 #include <pybind11/stl_bind.h>
 
 using namespace colmap;
+namespace py = pybind11;
 
 namespace PYBIND11_NAMESPACE {
 namespace detail {
 
+#include <limits>
+#include <stdexcept>
+
+#include <pybind11/pybind11.h>
+#include <pybind11/pytypes.h>
+
+// Custom type_caster from int to C++ uint types to support -1 as input.
+// Necessary for pycolmap id types that use -1 as the invalid value.
+#define TYPE_CASTER_UINT_T(DTYPE)                                         \
+  template <>                                                             \
+  struct type_caster<DTYPE> {                                             \
+   public:                                                                \
+    PYBIND11_TYPE_CASTER(DTYPE, _(#DTYPE));                               \
+    bool load(handle src, bool) {                                         \
+      if (!py::isinstance<py::int_>(src)) {                               \
+        return false;                                                     \
+      }                                                                   \
+      int64_t val = py::cast<int64_t>(src);                               \
+      if (val == -1) {                                                    \
```

Comment:
I don't like this approach because it will accept -1 as valid input for any function that accepts uint\*_t, while this is valid only for ID setters. Instead, how about wrapping `Set{Image,Camera,...}Id` with a templated function that accepts an int and casts appropriately?

Reply:
The use of pycolmap ids actually happen in a lot of places in the bindings (not only for the setters). For example, one can do ``grep -r '_id' ./src/pycolmap/`` and find a lot of relevant bindings. And the logic is also not consistent (e.g., here there is explicit casting supported for only one method: https://github.com/colmap/colmap/blob/main/src/pycolmap/sfm/incremental_mapper.cc#L358).

Actually I have a different proposal: how about we not supporting -1 as input and always recommend users to use ctypes to convert themselves from the Python side, as mentioned in the issue:
```
import pycolmap
import ctypes
p = pycolmap.Point2D()
p.point3D_id = ctypes.c_uint32(-1).value
```
We could update the logs here to inform the users who use -1 as input: https://github.com/B1ueber2y/colmap/blob/fix/pycolmap_uint/src/pycolmap/pybind11_extension.h#L44.

Reply:
For example, something like: https://github.com/B1ueber2y/colmap/commit/1639ddb4b3b620386f822b1618f15bd3476d1a39

Reply:
> The use of pycolmap ids actually happen in a lot of places in the bindings (not only for the setters). For example, one can do `grep -r '_id' ./src/pycolmap/` and find a lot of relevant bindings. And the logic is also not consistent (e.g., here there is explicit casting supported for only one method: https://github.com/colmap/colmap/blob/main/src/pycolmap/sfm/incremental_mapper.cc#L358).
> 
> Actually I have a different proposal: how about we not supporting -1 as input and always recommend users to use ctypes to convert from the Python side themselves, as the solution I suggested in the issue:
> 
> ```
> import pycolmap
> import ctypes
> p = pycolmap.Point2D()
> p.point3D_id = ctypes.c_uint32(-1).value
> ```
> 
> We could update the error logs here to inform the users who use -1 as input: https://github.com/B1ueber2y/colmap/blob/fix/pycolmap_uint/src/pycolmap/pybind11_extension.h#L44.

Update: this is not the right approach, since point3D_id is uint64_t, and ``p.point3D_id = ctypes.c_uint32(-1).value`` does not work with the following error: 
```
RuntimeError: Unable to cast Python instance of type <class 'int'> to C++ type '?' (#define PYBIND11_DETAILED_ERROR_MESSAGES or compile in debug mode for details)
```
@sarlinpe I agree that wrapping all functions with explicit appropriate casting is the best approach, though it will require going through all the relevant functions to fully fix the issue. Feel free to take this feature and go ahead : )

Reply:
Another idea is to support type_caster for ``ctypes.c_ulong`` and ``c_types.c_uint`` and always suggest the users to do: ``p.point3D_id = ctypes.c_uint64(-1)``. This may be a cleaner alternative.